### PR TITLE
Added else if Conditional Statements to the Metal backend #34

### LINKED ISSUE #34: Add `else if` Conditional Statements to the Metal backend
To add `else if` support to CrossGL, modify the following files in the `crosstl/src/backend/Metal` directory:

- [ ] **MetalLexer.py** : Add token recognition for `else if`.
- [ ] **MetalParser.py** : Update grammar and AST to handle `else if`.
- [ ] **MetalCrossGLCodeGen.py** : Implement `else if` for Crossgl output.

Your contribution is appreciated! Feel free to reach out on our [Discord Server ](https://discord.com/invite/uyRQKXhcyW) if you have any query.

### PR DESCRIPTION
**Pull request by adityanarayanan343**
**Added Support for else if in MetalParser**

1. **Grammar Update**: Enhanced the parse_if_statement method to handle **else if** constructs in Metal syntax.

2. **AST Update**: Modified the Abstract Syntax Tree (AST) structure to support else if nodes, ensuring correct hierarchical representation of nested conditional logic.

3. **Code Refactor**: Refactored the parse_if_statement method to improve readability and maintainability while integrating else if logic.

Closes #34

### CODE REVIEW

**Comment by adityanarayanan343** [comment]
Can u please check what else is there to change in this code?

**Review by samthakur587 — COMMENTED** [review]
thanks @adityanarayanan343  for contributing here i'll happy to solve any query if you in understanding the code.

**Comment by adityanarayanan343** [comment]
Can you check whether all the fixes you mentioned were addressed and please do tell if there were any other issues.

**Comment by samthakur587** [comment]
> Can you check whether all the fixes you mentioned were addressed and please do tell if there were any other issues.

Hii @adityanarayanan343 sorry for late reply your changes are looking great 👍 

One last thing can you please add a test for your else_if condition at metal backend you can find test files in tests folder. Before that pull the latest changes. 😄 

Thanks 
Sam

**Review by samthakur587 — COMMENTED** [review]
LGTM 🚀  waiting for tests to be added

**Comment by adityanarayanan343** [comment]
Sure thing @samthakur587 I will do it tomorrow morning and create a pull request.

**Comment by samthakur587** [comment]
> Sure thing @samthakur587 I will do it tomorrow morning and create a pull request.

You can write all tests in this pr also just pull latest changes add test and push on this branch.

**Comment by adityanarayanan343** [comment]
Hi @samthakur587 I have done testing and added if-

else test case to the test codes please check and comment on any changes. 
P.S: Attached Screenshot of getting all test cases passed

**Comment by NripeshN** [comment]
Hi @adityanarayanan343 
The test you added seems to [fail](https://github.com/CrossGL/crosstl/actions/runs/10501415601/job/29091365698?pr=39#step:5:105). Could you please look into it?

**Comment by samthakur587** [comment]
hii @adityanarayanan343  the tests are passing at metal codegen can you please have a look on this 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <crosstl.src.backend.Metal.MetalCrossGLCodeGen.MetalToCrossGLConverter object at 0x7f10543909a0>
node = IfNode(condition=BinaryOpNode(left=MemberAccessNode(object=MemberAccessNode(object=input, member=position), member=x),...right=VectorConstructorNode(type_name=float2, args=['0.5', '0.5']))], else_body=None)], elif_bodies=[], else_body=None)
indent = 3, is_main = True

    def generate_if_statement(self, node, indent, is_main):
        condition = self.generate_expression(node.condition, is_main)
    
        code = f"if ({condition}) {{\n"
        code += self.generate_function_body(node.if_body, indent + 1, is_main)
        code += "    " * indent + "}"
        for elif_node in node.elif_conditions:
            elif_condition = self.generate_expression(elif_node.condition, is_main)
            code += f" else if ({elif_condition}) {{\n"
>           code += self.generate_function_body(elif_node.body, indent + 1, is_main)
E           AttributeError: 'IfNode' object has no attribute 'body'

crosstl/src/backend/Metal/MetalCrossGLCodeGen.py:178: AttributeError
=========================== short test summary info ============================
FAILED tests/test_backend/test_metal/test_codegen.py::test_else_if - AttributeError: 'IfNode' object has no attribute 'body'
========================= 1 failed, 16 passed in 0.19s =========================

**Comment by adityanarayanan343** [comment]
Hi @samthakur587 ,

I think I have fixed the issue so can you please confirm the same thing and do comment on it.

Kudos,
Aditya

**Comment by samthakur587** [comment]
hii @adityanarayanan343  your tests are passing but not translating the correctly 

```
shader main {
 
    // Vertex Shader
    vertex {
        input vec3 position;
        output vec2 vUV;
        void main() {
                        gl_Position = vec4(position, 1.0);
            if (position.x == position.y) {
                vUV = vec2(0.0, 0.0);
            } else if ([]) {
                vUV = vec2(1.0, 1.0);
            }
                    }
    }

    // Fragment Shader
    fragment {
        input vec2 vUV;
        output vec4 fragColor;
        void main() {
                        if (vUV.x == vUV.y) {
                fragColor = vec4(0.0, 1.0, 0.0, 1.0);
            } else if ([]) {
                fragColor = vec4(1.0, 0.0, 0.0, 1.0);
            }
                    }
    }
}
```

in else if condition the only `[ ]` is translated . can you please have a look on this . and can you please write a separate test for else_if not in if condition.

**Comment by adityanarayanan343** [comment]
Hi @vaatsalya123  can you check whether this issue has been fixed or not and can you please comment on it.

Thanks,
Aditya

**Comment by samthakur587** [comment]
hii @adityanarayanan343  if you run this command .

```bash
pytest -v tests/test_backend/test_metal/test_codegen.py -s
```
then you can see the generated code for your test and if you see for else_if test then in this the condition is generated empty .

File: crosstl/src/backend/Metal/MetalParser.py [hunk 1]
```diff
@@ -306,15 +306,27 @@ def parse_variable_declaration_or_assignment(self):
 
     def parse_if_statement(self):
         self.eat("IF")
-        self.eat("LPAREN")
         condition = self.parse_expression()
-        self.eat("RPAREN")
         if_body = self.parse_block()
+
+        # Handle `else if`
+        elif_conditions = []
```
[the remark this answers is not in the document]
Comment (adityanarayanan343) [reply crosstl/src/backend/Metal/MetalParser.py:313]: Hi @samthakur587, I think I have added the required changes please do check on it and comment on it.

**Review by samthakur587 — APPROVED** [review]
hii @adityanarayanan343 great work i have fixed some small issues . now everything LGTM 🚀

**Pull request merged**